Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,29 +45,97 @@ void RebinToWorkspace::init() {
                   "Workspace2D histogram.");
 }
 
+namespace {
+bool needToRebin(const MatrixWorkspace_sptr &left,
+                 const MatrixWorkspace_sptr &rght) {
+  // if pointers match they are the same object
+  if (left == rght)
+    return false;
+
+  // look for first non-equal x-axis between the workspaces
+  const size_t numHist = left->getNumberHistograms();
+  for (size_t i = 0; i < numHist; ++i) {
+    if (left->getSpectrum(i).readX() != rght->getSpectrum(i).readX()) {
+      return true;
+    }
+  }
+  // everything must be the same
+  return false;
+}
+} // namespace
+
 /**
  * Execute the algorithm
  */
 void RebinToWorkspace::exec() {
   // The input workspaces ...
   MatrixWorkspace_sptr toRebin = getProperty("WorkspaceToRebin");
   MatrixWorkspace_sptr toMatch = getProperty("WorkspaceToMatch");
-  bool PreserveEvents = getProperty("PreserveEvents");
-
-  // First we need to create the parameter vector from the workspace with which
-  // we are matching
-  std::vector<double> rb_params = createRebinParameters(toMatch);
-
-  IAlgorithm_sptr runRebin = createChildAlgorithm("Rebin");
-  runRebin->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
-  runRebin->setPropertyValue("OutputWorkspace", "rebin_out");
-  runRebin->setProperty("params", rb_params);
-  runRebin->setProperty("PreserveEvents", PreserveEvents);
-  runRebin->setProperty("IgnoreBinErrors", true);
-  runRebin->executeAsChildAlg();
-  progress(1);
-  MatrixWorkspace_sptr ws = runRebin->getProperty("OutputWorkspace");
-  setProperty("OutputWorkspace", ws);
+  bool preserveEvents = getProperty("PreserveEvents");
+
+  if (needToRebin(toRebin, toMatch)) {
+    // TODO should copy code from Rebin to do the right thing spectrum by
+    // spectrum
+    if (!WorkspaceHelpers::commonBoundaries(*toMatch)) {
+      throw std::runtime_error(
+          "WorkspaceToMatch must have common bin boundaries in all spectra");
```

Comment:
Why don't we support this? It seems it would almost be easier to add support than handling the special case of not doing so? I see this is already in the old code, but since you are doing some cleanup anyway...

Using `createRebinParameters` and then calling `Rebin` is a kind of convoluted way? Why not simply do something like
```cpp
const auto edges = toRebin->binEdges(0);
for(size_t i=0; i<toRebin.getNumberHistograms(); ++i)
  toRebin->setHistogram(i, HistogramData::rebin(toRebin->histogram(i), edges));
```
If the workspace sizes match you could even use the correct edges for each spectrum.